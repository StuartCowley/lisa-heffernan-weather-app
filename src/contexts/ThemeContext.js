import React, { createContext, useContext, useState } from "react";
import PropTypes from "prop-types";

const ThemeContext = createContext();

export const useThemeContext = () => useContext(ThemeContext);

export const ThemeContextProvider = ({ children }) => {
  const [useDarkTheme, setUseDarkTheme] = useState(false);
  return (
    <ThemeContext.Provider value={{ useDarkTheme, setUseDarkTheme }}>
      {children}
    </ThemeContext.Provider>
  );
};

ThemeContextProvider.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
};
