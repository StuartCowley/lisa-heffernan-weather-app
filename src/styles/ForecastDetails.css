.forecast-details {
  margin: 4px auto 8px auto;  
  width: fit-content;
  padding: 8px;
  border-style: double;
  border-radius: 4px;
  border-width: 6px;
  text-align: center;
}

.forecast-details.light {
  background-color: white;
  color: black;
  border-color: black;   
}

.forecast-details.dark {
  background-color: black;
  color: white;
  border-color: white;
}
