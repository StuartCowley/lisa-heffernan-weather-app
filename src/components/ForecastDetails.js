import React from "react";
import PropTypes from "prop-types";
import {
  BsArrowUpCircle,
  BsArrowUpRightCircle,
  BsArrowRightCircle,
  BsArrowDownRightCircle,
  BsArrowDownCircle,
  BsArrowDownLeftCircle,
  BsArrowLeftCircle,
  BsArrowUpLeftCircle,
} from 'react-icons/bs';
import "../styles/ForecastDetails.css";
import { useThemeContext } from "../contexts/ThemeContext";

const ForecastDetails = ({ forecast }) => {
  const {
    date, temperature, humidity, wind,
  } = forecast;
  const { useDarkTheme } = useThemeContext();
  const formattedDate = new Date(date).toDateString();
  const getWindDirection = (windDirection) => {
    if (windDirection === 's' || windDirection === 'sse') {
      return <BsArrowUpCircle />;
    }
    if (windDirection === 'sw' || windDirection === 'ssw') {
      return <BsArrowUpRightCircle />;
    }
    if (windDirection === 'w' || windDirection === 'wsw') {
      return <BsArrowRightCircle />;
    }
    if (windDirection === 'nw' || windDirection === 'wnw') {
      return <BsArrowDownRightCircle />;
    }
    if (windDirection === 'n' || windDirection === 'nnw') {
      return <BsArrowDownCircle />;
    }
    if (windDirection === 'ne' || windDirection === 'nne') {
      return <BsArrowDownLeftCircle />;
    }
    if (windDirection === 'e' || windDirection === 'ene') {
      return <BsArrowLeftCircle />;
    }
    if (windDirection === 'se' || windDirection === 'ese') {
      return <BsArrowUpLeftCircle />;
    }
  };

  return (
    <div className={`forecast-details ${useDarkTheme ? "dark" : "light"}`}>
      <div className="forecast-details__date">{formattedDate}</div>
      <div className="forecast-details__max-temperature">
        Max Temperature:
        &nbsp;
        {temperature.max}
        &deg;C
      </div>
      <div className="forecast-details__min-temperature">
        Min Temperature:
        &nbsp;
        {temperature.min}
        &deg;C
      </div>
      <div className="forecast-details__humidity">
        Humidity:
        &nbsp;
        {humidity}
        %
      </div>
      <div className="forecast-details__wind" data-testid="wind-icon">
        Wind:
        &nbsp;
        {wind.speed}
        mph
        &nbsp;
        {getWindDirection(wind.direction)}
      </div>
    </div>
  );
};

ForecastDetails.propTypes = {
  forecast: PropTypes.shape({
    date: PropTypes.number,
    humidity: PropTypes.number,
    temperature: PropTypes.shape({
      max: PropTypes.number,
      min: PropTypes.number,
    }),
    wind: PropTypes.shape({
      speed: PropTypes.number,
      direction: PropTypes.string,
    }),
  }).isRequired,
};

export default ForecastDetails;
