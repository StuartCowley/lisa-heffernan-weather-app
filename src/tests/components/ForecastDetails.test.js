import React from "react";
import { render } from "@testing-library/react";
import ForecastDetails from "../../components/ForecastDetails";
import { ThemeContextProvider } from "../../contexts/ThemeContext";

describe("ForecastDetails", () => {
  const validProps = {
    date: 1111111,
    temperature: {
      max: 22,
      min: 12,
    },
    wind: {
      speed: 13,
      direction: "n",
    },
    humidity: 30,
  };

  it("renders correctly", () => {
    const { asFragment } = render(
      <ThemeContextProvider>
        <ForecastDetails forecast={validProps} />
      </ThemeContextProvider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders correct values for props", () => {
    const { getByText, getByTestId } = render(
      <ThemeContextProvider>
        <ForecastDetails forecast={validProps} />
      </ThemeContextProvider>,
    );

    expect(getByText("Thu Jan 01 1970")).toHaveClass("forecast-details__date");
    expect(getByText("Max Temperature: 22°C")).toHaveClass("forecast-details__max-temperature");
    expect(getByText("Min Temperature: 12°C")).toHaveClass("forecast-details__min-temperature");
    expect(getByText("Wind: 13mph")).toHaveClass("forecast-details__wind");
    expect(getByTestId("wind-icon")).toHaveClass("forecast-details__wind");
    expect(getByText("Humidity: 30%")).toHaveClass("forecast-details__humidity");
  });
});
