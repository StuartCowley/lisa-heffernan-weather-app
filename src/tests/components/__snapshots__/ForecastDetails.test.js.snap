// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForecastDetails renders correctly 1`] = `
<DocumentFragment>
  <div
    class="forecast-details light"
  >
    <div
      class="forecast-details__date"
    >
      Thu Jan 01 1970
    </div>
    <div
      class="forecast-details__max-temperature"
    >
      Max Temperature:  22°C
    </div>
    <div
      class="forecast-details__min-temperature"
    >
      Min Temperature:  12°C
    </div>
    <div
      class="forecast-details__humidity"
    >
      Humidity:  30%
    </div>
    <div
      class="forecast-details__wind"
      data-testid="wind-icon"
    >
      Wind:  13mph  
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M1 8a7 7 0 1 0 14 0A7 7 0 0 0 1 8zm15 0A8 8 0 1 1 0 8a8 8 0 0 1 16 0zM8.5 4.5a.5.5 0 0 0-1 0v5.793L5.354 8.146a.5.5 0 1 0-.708.708l3 3a.5.5 0 0 0 .708 0l3-3a.5.5 0 0 0-.708-.708L8.5 10.293V4.5z"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
